python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,510 @@
+        if value is None:
+            return "null"
+
+        if isinstance(self._type, FixedType):
+            return str(base64.b64encode(value))
+
+        return Identity._HUMAN_STRING_MAP.get(self._type, str)(value)
+
+
+class Truncate(Transform):
+    _VALID_TYPES = {IntegerType, LongType, StringType, BinaryType}
+
+    def __init__(self, transform_type: Type, width: int):
+        if transform_type not in Truncate._VALID_TYPES and not isinstance(
+            transform_type, DecimalType
+        ):
+            raise ValueError(f"Cannot truncate type: {transform_type}")
+
+        super().__init__(
+            f"truncate[{width}]",
+            f"Truncate(transform_type={repr(transform_type)}, width={width})",
+        )
+        self._type = transform_type
+        self._width = width
+
+    @property
+    def width(self):
+        return self._width
+
+    def apply(self, value):
+        if value is None:
+            return None
+        if self._type == IntegerType or self._type == LongType:
+            return value - (((value % self._width) + self._width) % self._width)
+        elif self._type == StringType or self._type == BinaryType:
+            return value[0 : min(self._width, len(value))]
+        else:  # decimal case
+            unscaled_value = Transform._unscale_decimal(value)
+            applied_value = unscaled_value - (
+                ((unscaled_value % self._width) + self._width) % self._width
+            )
+            return Decimal(f"{applied_value}e{value.as_tuple().exponent}")
+
+    def can_transform(self, target: Type) -> bool:
+        return self._type == target
+
+    def preserve_order(self) -> bool:
+        return True
+
+    def satisfy_order(self, other: Transform) -> bool:
+        if self._type == other:
+            return True
+        elif (
+            StringType == self._type
+            and isinstance(other, Truncate)
+            and StringType == other._type
+        ):
+            return self._width >= other._width
+
+        return False
+
+    def to_human_string(self, value) -> str:
+        if value is None:
+            return "null"
+
+        return str(base64.b64encode(value)) if self._type == BinaryType else str(value)
+
+
+class UnknownTransform(Transform):
+    def __init__(self, transform_type: Type, transform: str):
+        super().__init__(
+            transform,
+            f"UnknownTransform(transform_type={repr(transform_type)}, transform={transform})",
+        )
+        self._type = transform_type
+        self._transform = transform
+
+    def apply(self, value):
+        raise AttributeError(f"Cannot apply unsupported transform: {self.__str__()}")
+
+    def can_transform(self, target: Type) -> bool:
+        return self._type == target
+
+    def get_result_type(self, source: Type) -> Type:
+        return StringType
+
+
+VoidTransform = Transform("void", "VoidTransform")
+
+_HAS_WIDTH = re.compile("(\\w+)\\[(\\d+)\\]")
```

Comment:
Can you help me understand this regex and what it's used for?

Reply:
When parsing transform strings, truncate and bucket are parameterized, like `truncate[6]` and `bucket[256]`.

Reply:
`\\]` might not needed but we use the same one as Java one to keep it consistent.